R-package/R/callbacks.R

```diff
@@ -1,563 +1,923 @@
+#' the number of rounds to zero in order to overwrite rounds).
 #'
-#' The finalizer callback (called with \code{finalize = TURE} in the end) converts
-#' the \code{evaluation_log} list into a final data.table.
+#' Note that any R attributes that are assigned to the booster during the callback functions,
+#' will not be kept thereafter as the booster object variable is not re-assigned during
+#' training. It is however possible to set C-level attributes of the booster through
+#' \link{xgb.attr} or \link{xgb.attributes}, which should remain available for the rest
+#' of the iterations and after the training is done.
 #'
-#' The iteration evaluation result \code{bst_evaluation} must be a named numeric vector.
+#' For keeping variables across iterations, it's recommended to use `env` instead.
+#' \item data The data to which the model is being fit, as an `xgb.DMatrix` object.
 #'
-#' Note: in the column names of the final data.table, the dash '-' character is replaced with
-#' the underscore '_' in order to make the column names more like regular R identifiers.
+#' Note that, for \link{xgb.cv}, this will be the full data, while data for the specific
+#' folds can be found in the `model` object.
+#'
+#' \item watchlist The evaluation watchlist, as passed under argument `watchlist` to
+#' \link{xgb.train}.
+#'
+#' For \link{xgb.cv}, this will always be `NULL`.
+#'
+#' \item begin_iteration Index of the first boosting iteration that will be executed
+#' (base-1 indexing).
 #'
-#' Callback function expects the following values to be set in its calling frame:
-#' \code{evaluation_log},
-#' \code{bst_evaluation},
-#' \code{iteration}.
+#' This will typically be '1', but when using training continuation, depending on the
+#' parameters for updates, boosting rounds will be continued from where the previous
+#' model ended, in which case this will be larger than 1.
 #'
-#' @seealso
-#' \code{\link{callbacks}}
+#' \item end_iteration Index of the last boostign iteration that will be executed
+#' (base-1 indexing, inclusive of this end).
 #'
+#' It should match with argument `nrounds` passed to \link{xgb.train} or \link{xgb.cv}.
+#'
+#' Note that boosting might be interrupted before reaching this last iteration, for
+#' example by using the early stopping callback \link{xgb.cb.early.stop}.
+#'
+#' \item iteration Index of the iteration number that is being executed (first iteration
+#' will be the same as parameter `begin_iteration`, then next one will add +1, and so on).
+#'
+#' \item iter_feval Evaluation metrics for the `watchlist` that was supplied, either
+#' determined by the objective, or by parameter `feval`.
+#'
+#' For \link{xgb.train}, this will be a named vector with one entry per element in
+#' `watchlist`, where the names are determined as 'watchlist name' + '-' + 'metric name' - for
+#' example, if `watchlist` contains an entry named "tr" and the metric is "rmse",
+#' this will be a one-element vector with name "tr-rmse".
```

Comment:
Can we pass the evaluation result as a proxy instead of the actual data to callbacks?

Reply:
It is passed. But the user might _also_ want to use the actual data in a custom callback.

Reply:
We will do it your way.

My concern is that, often time the real data is difficult to protect, and might not be fully available. Should we pass the DMatrix or the original data, before or after processing like converting to numeric? Also, what happens with distributed training?

That said, I think your use case makes sense and will proceed as it's.

Reply:
I don't think those should be concerns, because the user can call `getinfo` on the DMatrix to obtain the data in a standardized format and use that in the callbacks.

As far as the function interface is aware, any input data other than `y` already needs to be passed as DMatrix for these functions, and that's what functions are meant to work with.